src/std/text/json/output.ss

```diff
@@ -0,0 +1,202 @@
+               (when (eq? (string-ref str 0) #\.)
+                 (write-char #\0 output))
+               (write-string str output)
+               (when (eq? (string-ref str (##fx- (string-length str) 1)) #\.)
+                 (write-char #\0 output))))
+
+           (def (write-json-list obj output env)
+             (write-char #\[ output)
+             (let lp ((rest obj))
+               (match rest
+                 ([val]                 ; last one
+                  (write-json-object val output env)
+                  (write-char #\] output))
+                 ([val . rest]
+                  (write-json-object val output env)
+                  (write-char #\, output)
+                  (lp rest))
+                 ([]                    ; empty
+                  (write-char #\] output)))))
+
+           (def (write-json-vector obj output env)
+             (let (len (vector-length obj))
+               (if (##fxpositive? len)
+                 (let (last (##fx- len 1))
+                   (begin
+                     (write-char #\[ output)
+                     (let lp ((n 0))
+                       (if (##fx= n last)
+                         (begin
+                           (write-json-object (##vector-ref obj n) output env)
+                           (write-char #\] output))
+                         (begin
+                           (write-json-object (##vector-ref obj n) output env)
+                           (write-char #\, output)
+                           (lp (##fx+ n 1)))))))
+                 (write-string "[]" output))))
+
+           (def (write-json-alist alist output env)
+             (write-char #\{ output)
+             (let lp ((rest alist))
+               (match rest
+                 ([[key . val] . rest]
+                  (let (key (json-key-string key))
+                    (write-char #\" output)
+                    (write-string key output)
+                    (write-char #\" output)
+                    (write-char #\: output)
+                    (write-json-object val output env)
+                    (unless (null? rest) (write-char #\, output))
+                    (lp rest)))
+               ([]                      ; empty
+                (write-char #\} output)))))
+
+           (def (write-json-alist/sort alist output env)
+             (write-json-alist (json-sort-alist alist) output env))
+
+           (def (write-json-hash obj output env)
+             (def lst (hash->list obj))
+             (if (&env-sort-keys env)
+               (write-json-alist/sort lst output env)
+               (write-json-alist lst output env)))
+
+           (def (write-json-string obj output env)
+             (def escape
+               '((#\" . #\")
+                 (#\\ . #\\)
+                 (#\backspace . #\b)
+                 (#\newline . #\n)
+                 (#\page . #\f)
+                 (#\return . #\r)
+                 (#\tab . #\t)))
+
+             (def (printable-char? char)
```

Comment:
Move that out and anything that doesn't need to be triplicated?

Reply:
Maybe. It does get guaranteed inlining this way.

Reply:
this doesn't really get triplicated, it is a tiny function and a constant that are not referenced outside; they get inlined.